File: bruteforce/1360-되돌리기/1.py
import sys
input = lambda : sys.stdin.readline().rstrip()
n = int(input())
stack = []
ans = ""
for _ in range(n):
    op, ch, time = input().split()
    if (op == "type"):
        time = int(time)
        if (stack):
            string = stack[-1][1] + ch
            stack.append((time, string))
        else:
            string = ch  # 헷갈리기 때문에 변수설정
            stack.append((time, ch))
    else:
        time, mTime = int(time), int(ch) # 헷갈리기 때문에 변수설정
        undoTime = time - mTime  # 80
        if(stack):
            curTime = stack[-1][0]
            if(curTime < undoTime):
                continue
            curIdx = len(stack)-1
            while(curIdx >= 0):
                curTime = stack[curIdx][0]
                if(undoTime <= curTime): # 15 17
                    curIdx -= 1
                else: # 13 > 12
                    break
            if(curIdx < 0):
                stack.append((time, ""))
            else:
                stack.append((time, stack[curIdx][1]))

if(stack): print(stack[-1][1])
else: print("")

File: src/main/kotlin/baekjun/bruteforce/1360-되돌리기/1.kt
package baekjun.bruteforce.`1360-되돌리기`

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))
lateinit var graph: MutableList<MutableList<Int>>
lateinit var visited: BooleanArray
fun main() {
    val n = br.readLine().toInt()
    val stack: MutableList<List<String>> = mutableListOf<List<String>>().apply {
        repeat(n) {
            add(br.readLine().split(" "))
        }
    }
    var curTime = stack[stack.lastIndex][2].toInt()
    val ans = StringBuilder("")

    while (stack.isNotEmpty()) {
        val (op, char, t) = stack.removeLast()
        val time = t.toInt()
        if (curTime >= time) {
            if (op == "type") {
                ans.insert(0, char)
                curTime--
            } else {
                val undoTime = char.toInt() // 헷갈려서 변수 지정했어요
                curTime = time - undoTime - 1
            }
        } else continue
    }
    bw.write("$ans\n")
    bw.flush()
    bw.close()
    br.close()
}
